himbaechel/uarch/gatemate/pll.cc

```diff
@@ -0,0 +1,352 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2024  The Project Peppercorn Authors.
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <boost/algorithm/string.hpp>
+
+#include "pack.h"
+
+#define HIMBAECHEL_CONSTIDS "uarch/gatemate/constids.inc"
+#include "himbaechel_constids.h"
+
+namespace {
+USING_NEXTPNR_NAMESPACE;
+
+double calculate_delta_stepsize(double num)
+{
+    std::ostringstream out;
+    out << std::fixed << std::setprecision(4) << num;
+    std::string str = out.str();
+    switch (4 - (str.size() - str.find_last_not_of('0') - 1)) {
+    case 4:
+        return 0.0001;
+    case 3:
+        return 0.001;
+    case 2:
+        return 0.01;
+    case 1:
+        return 0.1;
+    default:
+        return 1.0;
+    }
+}
+
+int getDCO_optimized_value(int f_dco_min, int f_dco_max, double f_dco)
+{
+    /*
+    optimization parameters
+    DCO_OPT:
+        1: Optimized to lower DCO frequency.
+        2: Optimized to the middle of the DCO range (DEFAULT)
+        3: Optimized to upper DCO frequency.
+    */
+    const int DCO_OPT = 2;              // 1: to lower DCO, 2: to DCO middle range, 3: to upper DCO, default = 2
+    const int DCO_FREQ_OPT_FACTOR = 24; // proportional factor btw. frequency match and dco optimization
+
+    switch (DCO_OPT) {
+    case 1:
+        return round(abs(f_dco - f_dco_min)) * DCO_FREQ_OPT_FACTOR;
+    case 3:
+        return round(abs(f_dco - f_dco_max)) * DCO_FREQ_OPT_FACTOR;
+    default: {
+        int f_dco_middle = (f_dco_max + f_dco_min) / 2; // Middle of DCO Range in MHz
+        return round(abs(f_dco - f_dco_middle)) * DCO_FREQ_OPT_FACTOR;
+    }
+    }
+}
+
+void get_M1_M2(double f_core, double f_core_delta, PllCfgRecord &setting, double max_input_freq)
```

Comment:
has this code come from colognechip? are there any particular license issues attached to it?

Reply:
This is rewrite from their code, but with direct permission. They agreed on license